pkg/server/controllers.go

```diff
@@ -209,13 +211,41 @@ func (s *Server) installClusterController(clientConfig clientcmdapi.Config, serv
 		cluster.Server = hostURL.String()
 	}
 
+	c := s.cfg.ClusterControllerOptions.Complete(*kubeconfig, s.kcpSharedInformerFactory, s.apiextensionsSharedInformerFactory)
+	cluster, apiresource, err := c.New()
+	if err != nil {
+		return err
+	}
+
 	if err := server.AddPostStartHook("install-cluster-controller", func(context genericapiserver.PostStartHookContext) error {
 		if err := s.waitForCRDServer(context.StopCh); err != nil {
 			return err
 		}
 
-		ctx := goContext(context)
-		return s.cfg.ClusterControllerOptions.Complete(*kubeconfig, s.kcpSharedInformerFactory, s.apiextensionsSharedInformerFactory).Start(ctx)
+		// TODO(sttts): remove CRD creation from controller startup
+		requiredCrds := []metav1.GroupKind{
+			{Group: apiresourceapi.GroupName, Kind: "apiresourceimports"},
+			{Group: apiresourceapi.GroupName, Kind: "negotiatedapiresources"},
+			{Group: clusterapi.GroupName, Kind: "clusters"},
+		}
+		for _, contextName := range []string{"admin", "user"} {
+			logicalClusterConfig, err := clientcmd.NewNonInteractiveClientConfig(*kubeconfig, contextName, &clientcmd.ConfigOverrides{}, nil).ClientConfig()
+			if err != nil {
+				return err
+			}
+			crdClient := apiextensionsv1client.NewForConfigOrDie(logicalClusterConfig).CustomResourceDefinitions()
+			if err := config.BootstrapCustomResourceDefinitions(goContext(context), crdClient, requiredCrds); err != nil {
+				return err
+			}
+		}
+
+		s.kcpSharedInformerFactory.WaitForCacheSync(context.StopCh)
+		s.kubeSharedInformerFactory.WaitForCacheSync(context.StopCh)
+
+		go cluster.Start(goContext(context), c.NumThreads)
```

Comment:
(No action required) In the context of running these controllers in-process, would it make sense to package this controller initialization for reuse i.e. by test code vs just embedding it in the server? And maybe that suggests that responsibility for initialization should remain with the controller (i.e. separation of concerns)?